File: src/main/java/plazma/kernel/lib/task/TaskContext.java
/*
 * Copyright (C) 2012-2023 Oleh Hapon ohapon@users.sourceforge.net
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307, USA.
 * 
 * Oleh Hapon
 * Kyiv, UKRAINE
 * ohapon@users.sourceforge.net
 */

package plazma.kernel.lib.task;

import plazma.kernel.lib.sys.Parameter;
import plazma.kernel.lib.sys.Parameters;

public class TaskContext {
    
    private String taskName;
    
    private Parameters parameters;

    public TaskContext() {
        super();
    }

    public String getTaskName() {
        return taskName;
    }

    public void setTaskName(String taskName) {
        this.taskName = taskName;
    }

    public Parameters getParameters() {
        if (parameters == null) {
            parameters = new Parameters();
        }
        return parameters;
    }

    public void setParameters(Parameters parameters) {
        this.parameters = parameters;
    }
    
    
    ////

    public Parameter getFoundParameter(String name) {

        ///////////////////////////////////////////
        Parameter parameter = getParameters().getParameter(name);
        if (parameter == null) {
            return null;
        }

        // We can have a parameter in the list but this parameter can be not found
        if (!parameter.isFound()) {
            return null;
        }

        return parameter;
        ////////////////////////////////////////////
    }

    ////

    public boolean hasParameter(String name) {

        ////////////////////////////////////////////
        Parameter parameter = getFoundParameter(name);
        return parameter != null;
        ////////////////////////////////////////////

        //return getParameters().hasParameter(name);
    }

    public String getStringParameter(String name) {
        return getParameters().getStringParameter(name);
    }

    public String getStringParameter(String name, String defValue) {
        return getParameters().getStringParameter(name, defValue);
    }

    public int getIntParameter(String name) {
        return getParameters().getIntParameter(name);
    }

    public int getIntParameter(String name, int defValue) {
        return getParameters().getIntParameter(name, defValue);
    }

    public float getFloatParameter(String name) {
        return getParameters().getFloatParameter(name);
    }

    public float getFloatParameter(String name, float defValue) {
        return getParameters().getFloatParameter(name, defValue);
    }

    public boolean getBooleanParameter(String name) {
        return getParameters().getBooleanParameter(name);
    }

    public boolean getBooleanParameter(String name, boolean defValue) {
        return getParameters().getBooleanParameter(name, defValue);
    }

    public boolean getFlagParameter(String name) {

        ////////////////////////////////////////////
        Parameter parameter = getFoundParameter(name);
        if (parameter == null) {
            return false;
        }

        if (!parameter.hasValue() || parameter.getValue().isEmpty()) {
            return true; // option - only name
        }
        ////////////////////////////////////////////

        return getParameters().getBooleanParameter(name);
    }

    public boolean getFlagParameter(String name, boolean defValue) {

        ////////////////////////////////////////////
        Parameter parameter = getFoundParameter(name);
        if (parameter == null) {
            return defValue;
        }

        if (!parameter.hasValue() || parameter.getValue().isEmpty()) {
            return true; // option - only name
        }
        ////////////////////////////////////////////

        return getParameters().getBooleanParameter(name, defValue);
    }    

}

File: src/main/cpp/plazma/kernel/lib/task/TaskContext.cpp
#include <string>
#include "TaskContext.h"

namespace task {

    TaskContext::TaskContext() {
        parameters = nullptr;
    }

    TaskContext::~TaskContext() {
        if (parameters != nullptr) {
            delete parameters;
        }
    }

    std::string TaskContext::getTaskName() {
        return taskName;
    }

    void TaskContext::setTaskName(std::string& taskName) {
        this->taskName = taskName;
    }

    Parameters* TaskContext::getParameters() {
        if (parameters == nullptr) {
            parameters = new Parameters();
        }
        return parameters;
    }

    void TaskContext::setParameters(Parameters* parameters) {
        this->parameters = parameters;
    }

    ////

    Parameter* TaskContext::getFoundParameter(const std::string& name) {

        ///////////////////////////////////////////
        Parameter* parameter = getParameters()->getParameter(name);
        if (parameter == nullptr) {
            return nullptr;
        }

        // We can have a parameter in the list but this parameter can be not found
        if (!parameter->isFound()) {
            return nullptr;
        }

        return parameter;
        ////////////////////////////////////////////
    }

    ////

    bool TaskContext::hasParameter(const std::string& name) {

        ////////////////////////////////////////////
        Parameter* parameter = getFoundParameter(name);
        return parameter != nullptr;
        ////////////////////////////////////////////

        //return getParameters()->hasParameter(name);
    }

    std::string TaskContext::getStringParameter(const std::string& name) {
        return getParameters()->getStringParameter(name);
    }

    std::string TaskContext::getStringParameter(const std::string& name, std::string defValue) {
        return getParameters()->getStringParameter(name, defValue);
    }

    int TaskContext::getIntParameter(const std::string& name) {
        return getParameters()->getIntParameter(name);
    }

    int TaskContext::getIntParameter(const std::string& name, int defValue) {
        return getParameters()->getIntParameter(name, defValue);
    }

    float TaskContext::getFloatParameter(const std::string& name) {
        return getParameters()->getFloatParameter(name);
    }

    float TaskContext::getFloatParameter(const std::string& name, float defValue) {
        return getParameters()->getFloatParameter(name, defValue);
    }

    bool TaskContext::getBooleanParameter(const std::string& name) {
        return getParameters()->getBooleanParameter(name);
    }

    bool TaskContext::getBooleanParameter(const std::string& name, bool defValue) {
        return getParameters()->getBooleanParameter(name, defValue);
    }

    bool TaskContext::getFlagParameter(const std::string& name) {

        ////////////////////////////////////////////
        Parameter* parameter = getFoundParameter(name);
        if (parameter == nullptr) {
            return false;
        }

        if (!parameter->hasValue() || parameter->getValue().empty()) {
            return true; // option - only name
        }
        ////////////////////////////////////////////

        return getParameters()->getBooleanParameter(name);
    }

    bool TaskContext::getFlagParameter(const std::string& name, bool defValue) {

        ////////////////////////////////////////////
        Parameter* parameter = getFoundParameter(name);
        if (parameter == nullptr) {
            return defValue;
        }

        if (!parameter->hasValue() || parameter->getValue().empty()) {
            return true; // option - only name
        }
        ////////////////////////////////////////////

        return getParameters()->getBooleanParameter(name, defValue);
    }

}
